design-decisions/2018-09-03-Owise-encoding.md

```diff
@@ -0,0 +1,77 @@
+Owise encoding
+==============
+
+Background
+----------
+
+K allows defining functions with a default case.
+A generic function definition would have multiple cases defined like this
+(not using actual K syntax):
+
+```
+f(ti1(xi1, ..., xi_ji), ..., ti_k(xi1, ..., xi_ji))
+    = Ti(xi1, ..., xi_ji)
+    if Pi(xi1, ..., xi_ji)
+```
+
+and it would have one default (`owise`) case that looks like this:
+
+```
+f(t1(x1, ..., xj), ..., tk(x1, ..., xj))
+    = T(x1, ..., xj)
+    if P(x1, ..., xj) [owise]
+```
+
+Note that all the above variables have implicit universal quantifiers.
+
+Decision
+--------
+
+Encode the axiom as:
+
+```
+f(t1(x1, ..., xj), ..., tk(x1, ..., xj))
+    = T(x1, ..., xj)
+    if And(
+        P(x1, ..., xj)
+        Not(
+            Or{i = 1, n}(
+                exists xi1 . exist xi2 . ... exists xi_ji (
+                    And(
+                        And{p = 1, k}(
+                            Ceil(And(ti_p(xi1, ..., xi_ji), tp(x1, ..., xj)))
+                        )
+                        Pi(xi1, ..., xi_ji)
+                    )
+                )
+            )
+        )
+    )
+```
```

Comment:
I'm in favor of the encoding. It's correct. It works even for cases when `t_1,...,t_k` are not constructors, but any symbols.